// robinhood-node/publish/core/Order.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Request = require('./Request');

var _Request2 = _interopRequireDefault(_Request);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Order = function () {
  _createClass(Order, null, [{
    key: 'place',
    value: function place(params) {
      if (!params.symbol || params.type === 'limit' && !params.price || !params.quantity || !params.side) return false;

      var defaultParams = {
        type: 'market',
        time_in_force: 'gfd',
        trigger: 'immediate'
      };

      var request = Object.assign(defaultParams, params);
      return _Request2.default.postPersonal('orders', request);
    }
  }]);

  function Order(id) {
    _classCallCheck(this, Order);

    this.id = id;
  }

  _createClass(Order, [{
    key: 'cancel',
    value: function cancel(id) {
      var orderId = id || this.id;
      return _Request2.default.postPersonal('orders/' + orderId + '/cancel');
    }
  }, {
    key: 'status',
    value: function status(id) {
      var orderId = id || this.id;
      return _Request2.default.getPersonal('orders/' + orderId);
    }
  }, {
    key: 'recent',
    get: function get() {
      return _Request2.default.getPersonal('orders');
    }
  }]);

  return Order;
}();

exports.default = Order;
